text/3390-daddy-borrowck-little-dropck.md

```diff
@@ -0,0 +1,470 @@
+
+The second use-case for a stable `may_dangle` is the ability to have `Drop` for
+self-referential types. This doesn't come up too often, but:
+
+```rust
+struct Foo<'a> {
+  this: Cell<Option<&'a Foo<'a>>>
+}
+
+impl<'a: '!> Drop for Foo<'a> {
+  fn drop(&mut self) {
+    ...
+  }
+}
+```
+
+Without this proposal, such a struct cannot be dropped. (It is, in fact,
+possible to write such a `Drop` impl. It just errors at site-of-use.)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Rust functions generally have an implicit requirement that all lifetimes must
+outlive the function. This can be desugared by treating:
+
+```rust
+fn foo<'a>(x: &'a ()) {}
+```
+
+as:
+
+```rust
+fn foo<'a: 'fn>(x: &'a ()) {}
+```
+
+Likewise, types also come with an implicit `'self` lifetime:
+
+```rust
+struct Foo<'a>(&'a ());
+```
+
+is really:
+
+```rust
+struct Foo<'a: 'self>(&'a ());
+```
+
+The ability to opt-out of this bound is called "non-local lifetimes". They're
+lifetimes external or out-of-scope to a type or function.
+
+This out-of-scope-ness can be represented using blocks:
+
+```rust
+fn foo<'a: '!, 'b: '!, 'c>(arg1: &'a (), arg2: &'b (), arg3: &'c ()) {
+    function_body();
+}
+```
+
+becomes:
+
+```rust
+let arg1;
+let arg2;
+let arg3;
+{ // block of non-local lifetimes
+    // need to make sure these are non-'static
+    let temporary_arg1 = ();
+    let temporary_arg2 = ();
+    arg1 = &temporary_arg1;
+    arg2 = &temporary_arg2;
+}
+let temporary_arg3 = ();
+arg3 = &temporary_arg3;
+function_body();
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Syntax
+
+The core syntax for opting out of the implicit lifetime bounds is `'!` in bound
+position. This is further restricted to only being allowed in functions
+(including trait functions), data types (structs, enums), and impls.
+
+This is an opt-out bound: it is similar to `?Sized`.
+
+## Interaction with borrowck
+
+The non-local lifetime interacts with borrowck the exact same way as the block
+representation above shows, replicated below for contextualization:
+
+```rust
+fn foo<'a: '!, 'b: '!, 'c>(arg1: &'a (), arg2: &'b (), arg3: &'c ()) {
+    function_body();
+}
+```
+
+becomes:
+
+```rust
+let arg1;
+let arg2;
+let arg3;
+{ // block of non-local lifetimes
+    // need to make sure these are non-'static
+    let temporary_arg1 = ();
+    let temporary_arg2 = ();
+    arg1 = &temporary_arg1;
+    arg2 = &temporary_arg2;
+}
+let temporary_arg3 = ();
+arg3 = &temporary_arg3;
+function_body();
+```
+
+This effectively makes arg1 and arg2 unusable in the function body.
+
+## Interactions with references
+
+References have a few special properties:
+
+1. They participate in implicit lifetime bounds.
+2. They're generally required to be alive.
+
+This RFC relaxes those properties. For the first one, we have that `&'a &'b T`
+does no longer imply `'b: 'a` if `'b: '!`, and thus `&'b T: 'a` is also no
```

Comment:
If `'!` becomes the top type (compared to `'static`, the bottom type) this still holds. All lifetimes (including valid and invalid) live for `'!`, with the caveat that it is not dereferencable (without unsafe). It's like an epitaph.

This also gives a reason for the new borrowck rule: If it cannot be dereferenced, it does not borrow.

Reply:
iirc the intention is specifically that `'!` isn't the top type, because previous approaches using a top type didn't go well.

Reply:
Yeah it's explicitly not a lifetime, only an opt-out bound.

In particular, it is never actually empty.